### PR TITLE
Add `insertMany` function in LeanIMT

### LINKED ISSUE #81: Create a function to add many leaves to the LeanIMT.sol tree

### PR DESCRIPTION
**Pull request by vplasencia**
## Related Issue

Closes #81 

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

### CODE REVIEW

File: packages/imt.sol/contracts/internal/InternalLeanIMT.sol [hunk 1]
```diff
@@ -73,6 +73,132 @@ library InternalLeanIMT {
...
+        for (uint256 i = 0; i < leaves.length; ) {
+            if (leaves[i] >= SNARK_SCALAR_FIELD) {
+                revert LeafGreaterThanSnarkScalarField();
+            } else if (leaves[i] == 0) {
+                revert LeafCannotBeZero();
+            } else if (_has(self, leaves[i])) {
+                revert LeafAlreadyExists();
+            }
+
+            self.leaves[leaves[i]] = self.size + i;
```
Comment (cedoor) [inline packages/imt.sol/contracts/internal/InternalLeanIMT.sol:93]: Is `self.size + 1` always the same value here? Looks like its value is not incremented.

Comment (vplasencia) [reply packages/imt.sol/contracts/internal/InternalLeanIMT.sol:93]: Hey! It is `i` not `1`

Comment (cedoor) [reply packages/imt.sol/contracts/internal/InternalLeanIMT.sol:93]: > Hey! It is `i` not `1`

Right!

File: packages/imt.sol/contracts/internal/InternalLeanIMT.sol [hunk 2]
```diff
@@ -73,6 +73,132 @@ library InternalLeanIMT {
...
+                nextLevel[i] = parentNode;
+
+                unchecked {
+                    ++i;
+                }
+            }
+
+            // Update the `sideNodes` variable.
+            // If `currentLevelSize` is odd, the saved value will be the last value of the array
+            // if it is even and there are more than 1 elements in `levelTemp`, the saved value
```
Comment (cedoor) [inline packages/imt.sol/contracts/internal/InternalLeanIMT.sol:163]:
```suggestion
            // if it is even and there are more than 1 element in `currentLevel`, the saved value
```

File: packages/imt.sol/contracts/internal/InternalLeanIMT.sol [hunk 3]
```diff
@@ -73,6 +73,132 @@ library InternalLeanIMT {
...
+            if (currentLevelSize & 1 == 1) {
+                self.sideNodes[level] = currentLevel[currentLevel.length - 1];
+            } else if (currentLevel.length > 1) {
+                self.sideNodes[level] = currentLevel[currentLevel.length - 2];
+            }
+
+            currentLevelStartIndex = nextLevelStartIndex;
+
+            // Calculate the next startIndex value.
+            // It is the position of the paret node which is pos/2.
```
Comment (cedoor) [inline packages/imt.sol/contracts/internal/InternalLeanIMT.sol:176]:
```suggestion
            // Calculate the next level startIndex value.
            // It is the position of the parent node which is pos/2.
```

**Review by cedoor — CHANGES_REQUESTED** [review]
Apart from a few typos, the rest looks great to me!

Great job 🔥

**Review by 0xjei — APPROVED** [review]
Great work! Few nits on comments and naming but overall looks good. Thanks @vplasencia for your time and effort!

File: packages/imt.sol/contracts/internal/InternalLeanIMT.sol (hunk 1 above)
Comment (vplasencia) [reply packages/imt.sol/contracts/internal/InternalLeanIMT.sol:93]: Yes, @0xjei it is correct.

**Review by cedoor — APPROVED** [review]

**Pull request merged**